tensorflow/compiler/jit/cluster_scoping_pass.cc

```diff
@@ -0,0 +1,148 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/jit/cluster_scoping_pass.h"
+
+#include "absl/algorithm/container.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+
+#include "tensorflow/compiler/jit/defs.h"
+#include "tensorflow/compiler/jit/xla_cluster_util.h"
+#include "tensorflow/core/framework/node_def_util.h"
+#include "tensorflow/core/graph/algorithm.h"
+
+namespace tensorflow {
+namespace {
+
+class ClusterScopingPassImpl {
+ public:
+  ClusterScopingPassImpl(Graph* graph,
+                         OptimizerOptions::GlobalJitLevel global_jit_level)
+      : graph_(graph),
+        global_jit_level_(global_jit_level),
+        unique_scope_id_(0) {}
+
+  Status Run();
+
+ private:
+  Status ScopingForPipelineStages();
+
+  size_t GetUniqueScopeId() { return unique_scope_id_++; }
+
+  void AddScopeToAllPredecessors(Node* start);
+
+  void AddScopeToAllSuccessors(Node* start);
+
+ private:
+  Graph* graph_;
+  OptimizerOptions::GlobalJitLevel global_jit_level_;
+  size_t unique_scope_id_;
+};
+
+absl::optional<string> GetXlaAutoJitScope(Node* node) {
+  string scope;
+  if (GetNodeAttr(node->attrs(), kXlaAutoJitScopeAttr, &scope).ok()) {
+    return scope;
+  }
+
+  return absl::nullopt;
+}
+
+void SetXlaAutoJitScope(Node* node, StringPiece scope) {
+  node->AddAttr(kXlaAutoJitScopeAttr, scope);
+}
+
+// NB! We append a new scope as suffix to the XlaAutoJitScope attribute instead
+// of overriding the old value.  In this way, we respect the original scopes.
+// In other words, appending X to Y creates the conjunction of the scopes X
+// and Y (i.e, X & Y in effect).
+void AddOrAppendXlaAutoJitScope(Node* node, absl::string_view suffix) {
+  string updated_scope;
+  absl::optional<string> cur_scope = GetXlaAutoJitScope(node);
+  if (cur_scope == absl::nullopt) {
+    updated_scope = std::string(suffix);
+  } else {
+    updated_scope = absl::StrCat(cur_scope.value(), "&", suffix);
+  }
+  SetXlaAutoJitScope(node, updated_scope);
+}
+
+void ClusterScopingPassImpl::AddScopeToAllPredecessors(Node* start) {
```

Comment:
This should be called `AddScopeToAllTransitivePredecessors`.  Same for `AddScopeToAllSuccessors`.

Reply:
Agreed.